### ISSUE TITLE
7300+ xunit tests, 8 threads faster than 32 threads

### ISSUE THREAD

**Issue by qazq** [xunit/visualstudio.xunit#284]
I have a .NET Core 2.2 project and 7385 xunit tests. All the tests test by in-memory SQLite. My old pc runs all tests spend about 7 minutes. Recently, I bought a new PC and expected the tests can be done in 3 minutes, but the result disappoints me.

Test runs by command line `dotnet test`

```
<PackageReference Include="xunit" Version="2.4.1" />
<PackageReference Include="xunit.runner.visualstudio" Version="2.4.1">
```

## Old PC
CPU: Intel i7-8705G (4 cores, 8 threads)
RAM: 32G DDR4-2400

The default max parallel threads of xunit framework are the number of logical processors。

Time: 0:07:02.8353695

![image](https://user-images.githubusercontent.com/997727/128435994-7a8f5da4-5c93-4018-8d73-b32e0e0670f3.png)

All logical processors usage almost 100% during the test.

## New PC
CPU: AMD R9 5950X (16 cores, 32 threads)
RAM: 32G DDR4-3200

Time: 0:06:55.9828601

It doesn't make sense. The time is almost the same!

The CPU usage was around 25% during the test.

![image](https://user-images.githubusercontent.com/997727/128452086-9c0787a6-25cb-422e-897c-fb017dcd331a.png)

I use `[assembly: CollectionBehavior(MaxParallelThreads = 1)]` to control the max parallel threads and got the results:

| MaxParallelThreads | Duration        |
| ------------------ | --------------- |
| 1                  | 0:08:57.1356187 |
| 2                  | 0:05:47.6916381 |
| 4                  | 0:04:15.5609807 |
| 8                  | 0:03:54.857479  |
| 10                 | 0:04:04.9681384 |
| 12                 | 0:04:13.8324604 |
| 16                 | 0:04:48.7449147 |
| 24                 | 0:05:46.1222699 |
| 32                 | 0:06:55.9828601 |

I got the best performance when the threads are 8.

I have no idea what the bottleneck is. My tests don't have to access SSD, so the bottleneck should not be I/O.

What can I do to analyze this issue?

**Comment by bradwilson** [comment]
My initial suspicion was that some bookkeeping was happening that was expensive (the code that schedules the work on the threads is itself single threaded). However, you're completing ~ 17.5 tests/second, which isn't a particularly difficult load for that single-threaded dispatcher to handle.

The assertion tests that we have are about 900 tests, so about one order of magnitude smaller than your example (but large enough that it should offer a decent benchmark). On my 5900X (24 thread CPU):

| MaxParallelThreads | Duration (.NET Core) | Duration (.NET Framework) |
| ------------------ | -------------------- | ------------------------- |
| 1 | 0.642s | 0.533s |
| 2 | 0.395s | 0.314s |
| 4 | 0.265s | 0.204s |
| 8 | 0.247s | 0.166s |
| 10 | 0.241s | 0.170s |
| 12 | 0.262s | 0.180s |
| 16 | 0.266s | 0.182s |
| 24 | 0.278s | 0.185s |
| 32 | 0.281s | 0.332s |

There's definitely a case of diminishing returns here. These tests are written to be extremely fast and highly parallelizable, yet they stop being effectively parallelized past about 8. They are purely run in memory: no I/O of any kind here (they're basically the "purest" version of isolated tests I have available.)

That said, with 8 parallel threads I'm only able to hit about a peak 20% CPU utilization. The CPU utilization when running with 32 threads peaks at around 77%, but obviously that's all bookkeeping noise, because I'm spending twice as long running the actual tests (in the case of .NET Framework).

So, the reality is: you shouldn't necessarily expect to be able to maximize the CPU usage. Choosing max parallel threads that is appropriate for your project is a bit of an art and not so much a science, I'm afraid. Being able to run multiple test assemblies in parallel might help you reach maximum CPU utilization while also getting the fastest possible results, as although each assembly uses perhaps a core or two's worth of compute power, many assemblies may be able to optimally use them all.

**Comment by bradwilson** [comment]
It's also worth noting that the big jump from 24 to 32 is something I expected (albeit not quite that big), because I have a 24-core CPU. Having 32 active threads on those 24 CPU cores means a lot of context switching, whereas 24 (or fewer) threads on a 24 core CPU affords the ability to keep each thread running on a single core. The reality is messier than that, of course (you have other things to run in your OS besides these tests), but exceeding is almost always going to be much more expensive, unless you have a ton of I/O bound tests which spend the majority of their time waiting for non-computational items to finish.

It's also worth noting that I tried to run the tests with unlimited max threads, and the results were very similar to using 24 cores (0.275s for .NET Core, 0.188s for .NET Framework), which suggests that the built-in thread pool in .NET might be optimizing itself to run approximately one thread per CPU core by default.

**Comment by bradwilson** [comment]
If you can provide a repro project that has a very specific performance problem that can be identified and fixed, please provide it. Otherwise I will end up closing this as "by design".

**Comment by bradwilson** [comment]
Closing for lack of response. Please open a new issue with a repro for further consideration. Thanks!
